### ISSUE TITLE
Exporter maybe not work when convert to onnx or other format model

### ISSUE THREAD

**Issue by wuRDmemory** [ultralytics/ultralytics#19102]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

Hi, I find an interesting issue in export model to onnx or other format model.

Here is the case: 

- I want to convert pertains yolov11n.pt to onnx, but with some custom code in the headers.

- So I write some code in Detect head(use export and format variables), but it totally does not work.

- I debuged the code in exporter.py to see why it happened, and here is my debug code in exporter.py 321 line:

```python
        from ultralytics.nn.modules import Detect as ModuleDetect ## this is original code, but I have to rename it to test
        from ultralytics.nn.modules.head import Detect, RTDETRDecoder        

        for m in model.modules():
            if type(m) == Detect:
                print("model type: ", n, type(m), type(m).__name__)
                print("Debug info:")
                print(f"Module name: {n}")
                print(f"Module type: {type(m)}")
                print(f"Module type name: {type(m).__name__}")
                print(f"Detect class: {Detect}")
                print(f"ModuleDetect class: {ModuleDetect}")
                print(f"Is instance check: {isinstance(m, Detect)}")
                print(f"Is instance check2: {isinstance(m, ModuleDetect)}")
                print(f"Module class path: {m.__class__.__module__}")
                print(f"Detect class path: {Detect.__module__}")
                print(f"ModuleDetect class path: {ModuleDetect.__module__}")
                print(f"Module base classes: {m.__class__.__bases__}")
                print(f"Module Detect: {id(ModuleDetect)} vs {id(Detect)} vs {id(type(m))}")

            if isinstance(m, Classify):
                m.export = True
                print("m is classify", m.export, m.format)
            if isinstance(m, (Detect, RTDETRDecoder)):  # includes all Detect subclasses like Segment, Pose, OBB
                m.dynamic = self.args.dynamic
                m.export = True
                m.format = self.args.format
                m.max_det = self.args.max_det
                print("m is detect", m.export, m.format)
            elif isinstance(m, C2f) and not is_tf_format:
                # EdgeTPU does not support FlexSplitV while split provides cleaner ONNX graph
                m.forward = m.forward_split
            if isinstance(m, Detect) and imx:
                from ultralytics.utils.tal import make_anchors

                m.anchors, m.strides = (
                    x.transpose(0, 1)
                    for x in make_anchors(
                        torch.cat([s / m.stride.unsqueeze(-1) for s in self.imgsz], dim=1), m.stride, 0.5
                    )
                )
``` 

- output show below, you can see when you see the Detect class itself, the output all says <class 'ultralytics.nn.modules.head.Detect'>, but when you check when id or isinstance function, they are different.
```
Debug info:
Module name: model.23
Module type: <class 'ultralytics.nn.modules.head.Detect'>
Module type name: Detect
Detect class: <class 'ultralytics.nn.modules.head.Detect'>
ModuleDetect class: <class 'ultralytics.nn.modules.head.Detect'>
Is instance check: True
Is instance check2: False
Module class path: ultralytics.nn.modules.head
Detect class path: ultralytics.nn.modules.head
ModuleDetect class path: ultralytics.nn.modules.head
Module base classes: (<class 'torch.nn.modules.module.Module'>,)
Module Detect: 701843600 vs 588963328 vs 588963328
````

- I check all project file where you use Detect, all import from "ultralytics.nn.modules", so I test with parse model from yaml, all works well, does not have to import from "ultralytics.nn.modules.head"

- So, this maybe a very interesting bug when load model from checkpoint I guess?

Does anyone have any idea about this ??

### Environment

Ultralytics 8.3.59 🚀 Python-3.10.14 torch-1.11.0+cu113 CPU (AMD EPYC 7543 32-Core Processor)
Setup complete ✅ (64 CPUs, 251.6 GB RAM, 892.1/960.2 GB disk)

OS                  Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.31
Environment         Linux
Python              3.10.14
Install             git
RAM                 251.62 GB
Disk                892.1/960.2 GB
CPU                 AMD EPYC 7543 32-Core Processor
CPU count           64
GPU                 None
GPU count           None
CUDA                None

matplotlib          ✅ 3.8.4>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.13.0>=1.4.1
torch               ✅ 1.11.0+cu113>=1.8.0
torchvision         ✅ 0.12.0+cu113>=0.9.0
tqdm                ✅ 4.66.2>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.2.1>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
{'OS': 'Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.31',
 'Environment': 'Linux',
 'Python': '3.10.14',
 'Install': 'git',
 'RAM': '251.62 GB',
 'Disk': '892.1/960.2 GB',
 'CPU': 'AMD EPYC 7543 32-Core Processor',
 'CPU count': 64,
 'GPU': None,
 'GPU count': None,
 'CUDA': None,
 'Package Info': {'matplotlib': '✅ 3.8.4>=3.3.0',
  'opencv-python': '✅ 4.9.0.80>=4.6.0',
  'pillow': '✅ 10.3.0>=7.1.2',
  'pyyaml': '✅ 6.0.1>=5.3.1',
  'requests': '✅ 2.31.0>=2.23.0',
  'scipy': '✅ 1.13.0>=1.4.1',
  'torch': '✅ 1.11.0+cu113>=1.8.0',
  'torchvision': '✅ 0.12.0+cu113>=0.9.0',
  'tqdm': '✅ 4.66.2>=4.64.0',
  'psutil': '✅ 5.9.8',
  'py-cpuinfo': '✅ 9.0.0',
  'thop': '✅ 0.1.1-2209072238>=0.1.1',
  'pandas': '✅ 2.2.1>=1.1.4',
  'seaborn': '✅ 0.13.2>=0.11.0'}}

### Minimal Reproducible Example

```
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @wuRDmemory, thank you for your detailed and insightful bug report 🚀! We appreciate the effort you put into debugging and providing extensive context for your issue.

If this is indeed a 🐛 Bug Report, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) if not already available? This will help us further investigate and resolve the issue more efficiently.

In the meantime, we recommend ensuring you are running the latest version of Ultralytics by upgrading your environment as follows:

## Upgrade

Upgrade the `ultralytics` package and its [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) using the command:

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in any of the following officially supported environments with dependencies pre-installed:
- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: Follow the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/).
- **AWS**: See the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker**: Or use the official [Docker Image](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

## Community & Support

While an Ultralytics engineer will review and address your issue soon, don’t hesitate to explore other ways to engage:
- Check the [Ultralytics Docs](https://docs.ultralytics.com/) for helpful resources.
- Join discussions on our [Discord](https://discord.com/invite/ultralytics) 🎧, [Discourse](https://community.ultralytics.com/), or [Subreddit](https://reddit.com/r/Ultralytics).

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

This badge reflects the current pass state of our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule), which performs daily tests on macOS, Windows, and Ubuntu to ensure compatibility of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/).

Thank you for your patience and your contribution to improving the Ultralytics ecosystem! 😊

**Comment by Y-T-G** [comment]
The model loaded from `.pt` wouldn't use your custom `Detect`. It loads the model object from the `.pt`. It doesn't recreate the model.

If you want to recreate the model, then you will have to create one using yaml and then load the weights:

```python
feom ultralytics import YOLO

model = YOLO("yolo11n.pt")  # doesn't recreate the model. No changes in the layer code will be reflected.
model = YOLO("yolo11n.yaml").load("yolo11n.pt")  # recreates the model and transfers the weights.
```

**Comment by wuRDmemory** [comment]
@Y-T-G Thanks for your response, really make sense!
